### DISCUSSION TITLE
Signed s3 redirects and content security policies

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5995] [Q&A]
I am working on a JSVM hook to intercept `onFileDownloadRequest` and return a 302 to a signed s3 URL instead of proxying through the pocketbase instance. Once I got it working, I immediately hit [img-src Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/img-src) restrictions.

For example, the PocketBase admin has this meta tag:

```html
 <meta
        http-equiv="Content-Security-Policy"
        content="default-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' http://127.0.0.1:* data: blob:; connect-src 'self' http://127.0.0.1:*; script-src 'self' 'sha256-GRUzBA7PzKYug7pqxv5rJaec5bwDCw1Vo6/IXwvD3Tc='"
    />
```

Even if I return a `Content-Security-Policy` response header, the `<meta>` value overrides it. So there are at least some cases where it's not possible to respond with a 302 to a different domain because CSP will block it.

I'm curious if anyone has more thoughts about this. I think, at least for PocketHost, it could help performance a lot if we could respond with 302's rather than proxying all file data through the instance because then the payload will load from CDN instead of the instance origin. I am considering adding an `Optimize Image Loading` toggle in the PocketHost instance dashboard so users can disable optimization if they hit CSP's that they can't work around.

Related: #5983, #4260

**Comment by ganigeorgiev** [comment]
I'll consider with the next release to replace it with a header but for now you can ignore admins/superusers requests in the hook and fallback to the original behavior.

**Reply by benallfree** [reply]
Yes I'm ignoring based on referer `/_/*`. Seems to be working.

**Reply by ganigeorgiev** [reply]
_Just for info, in [v0.23.4](https://github.com/pocketbase/pocketbase/releases/tag/v0.23.4) I've changed the UI CSP to be served as response header so you should be able to overwrite it if a `Content-Security-Policy` header is set with a middleware before the route finalizer (if you are going to overwrite it I'd advice to make your changes based on the existing CSP value)._
